// sparql-engine/dist/engine/executors/graph-executor.js
/* file : graph-executor.ts
MIT License

Copyright (c) 2018 Thomas Minier

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var executor_1 = require("./executor");
var pipeline_1 = require("../pipeline/pipeline");
var utils_1 = require("../../utils");
/**
 * A GraphExecutor is responsible for evaluation a GRAPH clause in a SPARQL query.
 * @author Thomas Minier
 */
var GraphExecutor = /** @class */ (function (_super) {
    __extends(GraphExecutor, _super);
    /**
     * Constructor
     * @param dataset - RDF Dataset used during query execution
     */
    function GraphExecutor(dataset) {
        var _this = _super.call(this) || this;
        _this._dataset = dataset;
        return _this;
    }
    /**
     * Build a {@link PipelineStage} to evaluate a GRAPH clause
     * @param  source  - Input {@link PipelineStage}
     * @param  node    - Graph clause
     * @param  options - Execution options
     * @return A {@link PipelineStage} used to evaluate a GRAPH clause
     */
    GraphExecutor.prototype.buildIterator = function (source, node, context) {
        var _this = this;
        var subquery;
        if (node.patterns[0].type === 'query') {
            subquery = node.patterns[0];
        }
        else {
            subquery = {
                prefixes: context.getProperty('prefixes'),
                queryType: 'SELECT',
                variables: ['*'],
                type: 'query',
                where: node.patterns
            };
        }
        var engine = pipeline_1.Pipeline.getInstance();
        // handle the case where the GRAPh IRI is a SPARQL variable
        if (utils_1.rdf.isVariable(node.name) && context.namedGraphs.length > 0) {
            // clone the source first
            source = engine.clone(source);
            // execute the subquery using each graph, and bound the graph var to the graph iri
            var iterators = context.namedGraphs.map(function (iri) {
                return engine.map(_this._execute(source, iri, subquery, context), function (b) {
                    return b.extendMany([[node.name, iri]]);
                });
            });
            return engine.merge.apply(engine, __spread(iterators));
        }
        // otherwise, execute the subquery using the Graph
        return this._execute(source, node.name, subquery, context);
    };
    /**
     * Returns a {@link PipelineStage} used to evaluate a GRAPH clause
     * @param  source    - Input {@link PipelineStage}
     * @param  iri       - IRI of the GRAPH clause
     * @param  subquery  - Subquery to be evaluated
     * @param  options   - Execution options
     * @return A {@link PipelineStage} used to evaluate a GRAPH clause
     */
    GraphExecutor.prototype._execute = function (source, iri, subquery, context) {
        var opts = context.clone();
        opts.defaultGraphs = [iri];
        return this._builder._buildQueryPlan(subquery, opts, source);
    };
    return GraphExecutor;
}(executor_1.default));
exports.default = GraphExecutor;
